### PR TITLE
[data loader] make more iterator attributes private

### PR DESCRIPTION
**Pull request by ssnl**
1. Prefixed underscores to any `DataLoaderIter` attribute that is not part of the data loader ctor argument list. 
2. Prefixed `DataLoader.dataset_kind` with underscore because it only makes sense with the private enum `_DatasetKind`, and is an implementation detail.
3. Disallow setting `DataLoader.dataset` and `DataLoader.batch_sampler` after initializing a `DataLoader` because they affect other attributes in `__init__`.

These changes should not have major BC breaking effect since the big changes are on the iterator class and most users don't even store it. I GitHub searched `pin_memory_thread` and (while I didn't look through all result pages) results I see are forks of pytorch and blog posts on how data loader works.

### CODE REVIEW

File: torch/utils/data/dataloader.py [hunk 1]
```diff
@@ -300,28 +300,28 @@ def __len__(self):
...
         self.dataset = loader.dataset
-        self.dataset_kind = loader.dataset_kind
-        self.auto_collation = loader._auto_collation
+        self._dataset_kind = loader._dataset_kind
+        self._auto_collation = loader._auto_collation
         self.drop_last = loader.drop_last
-        self.index_sampler = loader._index_sampler
+        self._index_sampler = loader._index_sampler
         self.num_workers = loader.num_workers
         self.pin_memory = loader.pin_memory and torch.cuda.is_available()
```
Comment (apaszke) [inline torch/utils/data/dataloader.py]: FWIW changing `num_workers` or `pin_memory` might result in some pretty bad bugs too, so we might just as well make everything private 🤷‍♂

Comment (SsnL) [reply torch/utils/data/dataloader.py]: I made everything private :)

**Review by ezyang — APPROVED** [review]

**Comment by PetrochukM** [comment]
By the @SsnL, making `batch_sampler` private broke this sample code: https://github.com/pytorch/pytorch/issues/15849

How do you recommend working around this?

**Comment by PetrochukM** [comment]
@SsnL Checking back in!

**Pull request closed without merging**